### Feature Details:

`instance_id`: A unique identifier for each instance in the dataset, represented as a floating-point number. There are 50,000 instances in total

`artist_name`: The name of the artist who created the track, represented as a string (object).

`track_name`: The name of the track, represented as a string (object

`popularity`: A measure of the popularity of the track, represented as a floating-point number between 0 and 100. Higher values indicate greater popularity.

`acousticness`: A measure of the acousticness of the track, represented as a floating-point number between 0 and 1. Higher values indicate greater acousticness.

`danceability`: A measure of the danceability of the track, represented as a floating-point number between 0 and 1. Higher values indicate greater danceability.

`duration_ms`: The duration of the track in milliseconds, represented as a floating-point number.

`energy`: A measure of the energy of the track, represented as a floating-point number between 0 and 1. Higher values indicate greater energy.

`instrumentalness`: A measure of the instrumentalness of the track, represented as a floating-point number between 0 and 1. Higher values indicate greater instrumentalness.

`key`: The key in which the track is performed, represented as a string (object). There are 12 possible values: 'C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G', 'G#', 'A', 'A#', and 'B'.

`liveness`: A measure of the liveness of the track, represented as a floating-point number between 0 and 1. Higher values indicate greater liveness.

`loudness`: A measure of the loudness of the track, represented as a floating-point number in decibels (dB). Higher values indicate greater loudness.

`mode`: The mode in which the track is performed, represented as a string (object). There are two possible values: 'major' and 'minor'.

`speechiness`: A measure of the speechiness of the track, represented as a floating-point number between 0 and 1. Higher values indicate greater speechiness.

`tempo`: The tempo of the track in beats per minute (BPM), represented as a string (object).

`obtained_date`: The date on which the track information was obtained, represented as a string (object).

`valence`: A measure of the valence (positivity) of the track, represented as a floating-point number between 0 and 1. Higher values indicate greater positivity.

#### `music_genre`: The genre of the track, represented as a string (object). There are multiple possible genres in the dataset.

In [306]:
df.head()

instance_id           artist_name            track_name  popularity  \
0      32894.0              Röyksopp  Röyksopp's Night Out        27.0   
1      46652.0  Thievery Corporation      The Shining Path        31.0   
2      30097.0        Dillon Francis             Hurricane        28.0   
3      62177.0              Dubloadz                 Nitro        34.0   
4      24907.0           What So Not      Divide & Conquer        32.0   

   acousticness  danceability  duration_ms  energy  instrumentalness key  \
0       0.00468         0.652         -1.0   0.941           0.79200  A#   
1       0.01270         0.622     218293.0   0.890           0.95000   D   
2       0.00306         0.620     215613.0   0.755           0.01180  G#   
3       0.02540         0.774     166875.0   0.700           0.00253  C#   
4       0.00465         0.638     222369.0   0.587           0.90900  F#   

   liveness  loudness   mode  speechiness               tempo obtained_date  \
0     0.115    -5.201  Minor       0.0748             100.889         4-Apr   
1     0.124    -7.043  Minor       0.0300  115.00200000000001         4-Apr   
2     0.534    -4.617  Major       0.0345             127.994         4-Apr   
3     0.157    -4.498  Major       0.2390             128.014         4-Apr   
4     0.157    -6.266  Major       0.0413             145.036         4-Apr   

   valence music_genre  
0    0.759  Electronic  
1    0.531  Electronic  
2    0.333  Electronic  
3    0.270  Electronic  
4    0.323  Electronic

In [307]:
df.shape

(50005, 18)

In [308]:
df.columns

Index(['instance_id', 'artist_name', 'track_name', 'popularity',
       'acousticness', 'danceability', 'duration_ms', 'energy',
       'instrumentalness', 'key', 'liveness', 'loudness', 'mode',
       'speechiness', 'tempo', 'obtained_date', 'valence', 'music_genre'],
      dtype='object')

In [309]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50005 entries, 0 to 50004
Data columns (total 18 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   instance_id       50000 non-null  float64
 1   artist_name       50000 non-null  object 
 2   track_name        50000 non-null  object 
 3   popularity        50000 non-null  float64
 4   acousticness      50000 non-null  float64
 5   danceability      50000 non-null  float64
 6   duration_ms       50000 non-null  float64
 7   energy            50000 non-null  float64
 8   instrumentalness  50000 non-null  float64
 9   key               50000 non-null  object 
 10  liveness          50000 non-null  float64
 11  loudness          50000 non-null  float64
 12  mode              50000 non-null  object 
 13  speechiness       50000 non-null  float64
 14  tempo             50000 non-null  object 
 15  obtained_date     50000 non-null  object 
 16  valence           50000 non-null  float6

In [310]:
df.isnull().sum()

instance_id         5
artist_name         5
track_name          5
popularity          5
acousticness        5
danceability        5
duration_ms         5
energy              5
instrumentalness    5
key                 5
liveness            5
loudness            5
mode                5
speechiness         5
tempo               5
obtained_date       5
valence             5
music_genre         5
dtype: int64

In [311]:
df = df.drop_duplicates()

In [312]:
df.shape

(50001, 18)

In [313]:
df.isnull().sum()

instance_id         1
artist_name         1
track_name          1
popularity          1
acousticness        1
danceability        1
duration_ms         1
energy              1
instrumentalness    1
key                 1
liveness            1
loudness            1
mode                1
speechiness         1
tempo               1
obtained_date       1
valence             1
music_genre         1
dtype: int64

In [314]:
df.dropna(inplace=True)

In [315]:
df.shape

(50000, 18)

In [316]:
df.describe()

instance_id    popularity  acousticness  danceability   duration_ms  \
count  50000.000000  50000.000000  50000.000000  50000.000000  5.000000e+04   
mean   55888.396360     44.220420      0.306383      0.558241  2.212526e+05   
std    20725.256253     15.542008      0.341340      0.178632  1.286720e+05   
min    20002.000000      0.000000      0.000000      0.059600 -1.000000e+00   
25%    37973.500000     34.000000      0.020000      0.442000  1.748000e+05   
50%    55913.500000     45.000000      0.144000      0.568000  2.192810e+05   
75%    73863.250000     56.000000      0.552000      0.687000  2.686122e+05   
max    91759.000000     99.000000      0.996000      0.986000  4.830606e+06   

             energy  instrumentalness      liveness      loudness  \
count  50000.000000      50000.000000  50000.000000  50000.000000   
mean       0.599755          0.181601      0.193896     -9.133761   
std        0.264559          0.325409      0.161637      6.162990   
min        0.000792          0.000000      0.009670    -47.046000   
25%        0.433000          0.000000      0.096900    -10.860000   
50%        0.643000          0.000158      0.126000     -7.276500   
75%        0.815000          0.155000      0.244000     -5.173000   
max        0.999000          0.996000      1.000000      3.744000   

        speechiness       valence  
count  50000.000000  50000.000000  
mean       0.093586      0.456264  
std        0.101373      0.247119  
min        0.022300      0.000000  
25%        0.036100      0.257000  
50%        0.048900      0.448000  
75%        0.098525      0.648000  
max        0.942000      0.992000

##### let's drop the following columns which won't contribute to the model using domain knowledge.
1. instance_id
2. track_name
3. obtained_date
4. duration_ms

In [317]:
df.drop(["instance_id", "obtained_date", "track_name", "duration_ms"], axis=1, inplace=True)

Some names were not provided for some artists, which could have bettered performance i.e. empty_field. Hence we remove them

In [318]:
df['artist_name'].value_counts()

artist_name
empty_field                2489
Nobuo Uematsu               429
Wolfgang Amadeus Mozart     402
Ludwig van Beethoven        317
Johann Sebastian Bach       314
                           ... 
Ward Davis                    1
A.A. Bondy                    1
Iris DeMent                   1
Mac Davis                     1
Millonario                    1
Name: count, Length: 6863, dtype: int64

In [319]:
df = df[df['artist_name'] != 'empty_field']

In [320]:
# data variables are object ar boolean

cat_cols = list(df.dtypes[df.dtypes == 'object'].index)
print(f'Columns with categorical variables are: {cat_cols} \n')
print(f'Categorical Features: {len(cat_cols)} \n')


num_cols = list(df.dtypes[df.dtypes == 'float64'].index)
print(f'Columns with numerical values are: {num_cols} \n')
print(f'Numerical Features: {len(num_cols)}')

Columns with categorical variables are: ['artist_name', 'key', 'mode', 'tempo', 'music_genre'] 

Categorical Features: 5 

Columns with numerical values are: ['popularity', 'acousticness', 'danceability', 'energy', 'instrumentalness', 'liveness', 'loudness', 'speechiness', 'valence'] 

Numerical Features: 9


In [321]:
df.tempo.value_counts()

tempo
?                     4720
140.007                 16
120.0                   16
100.00299999999999      16
130.016                 15
                      ... 
144.249                  1
132.74                   1
86.073                   1
160.488                  1
99.20100000000001        1
Name: count, Length: 28373, dtype: int64

In [322]:
df[df["tempo"] == "?"]

artist_name  popularity  acousticness  danceability  energy  \
5        Axel Boman        47.0       0.00523         0.755   0.731   
32         PEEKABOO        45.0       0.02330         0.729   0.869   
35     Fabian Mazur        33.0       0.10800         0.493   0.682   
36       Wax Tailor        45.0       0.04780         0.646   0.649   
39             Dahu        37.0       0.20300         0.769   0.551   
...             ...         ...           ...           ...     ...   
49918      Big Sean        58.0       0.29600         0.379   0.644   
49964    Millonario        59.0       0.08470         0.929   0.737   
49967     MadeinTYO        62.0       0.17900         0.860   0.625   
49976      Big Sean        52.0       0.70000         0.462   0.741   
49977     Nate Dogg        58.0       0.10500         0.905   0.414   

       instrumentalness key  liveness  loudness   mode  speechiness tempo  \
5              0.854000   D    0.2160   -10.517  Minor       0.0412     ?   
32             0.585000   F    0.0944    -7.569  Minor       0.1290     ?   
35             0.000000   A    0.1960    -5.781  Minor       0.2870     ?   
36             0.002520   G    0.3530    -5.259  Major       0.1170     ?   
39             0.882000  A#    0.1090   -12.520  Minor       0.0408     ?   
...                 ...  ..       ...       ...    ...          ...   ...   
49918          0.000000  A#    0.3130    -7.933  Minor       0.1120     ?   
49964          0.000000  G#    0.8610    -6.424  Major       0.1020     ?   
49967          0.000136   D    0.3000    -6.922  Major       0.1050     ?   
49976          0.000000  A#    0.3400    -8.008  Minor       0.1620     ?   
49977          0.000366  G#    0.0914    -8.112  Minor       0.0615     ?   

       valence music_genre  
5       0.6140  Electronic  
32      0.1380  Electronic  
35      0.2390  Electronic  
36      0.7270  Electronic  
39      0.0618  Electronic  
...        ...         ...  
49918   0.1440     Hip-Hop  
49964   0.8860     Hip-Hop  
49967   0.2210     Hip-Hop  
49976   0.5890     Hip-Hop  
49977   0.7580     Hip-Hop  

[4720 rows x 14 columns]

##### Also, we need to drop "?" values in the tempo column since we cannot ascribe it with something else.
Then convert it to a float

In [323]:
df = df.drop(df[df["tempo"] == "?"].index)

df["tempo"] = df["tempo"].astype("float64")

In [324]:
df.dtypes

artist_name          object
popularity          float64
acousticness        float64
danceability        float64
energy              float64
instrumentalness    float64
key                  object
liveness            float64
loudness            float64
mode                 object
speechiness         float64
tempo               float64
valence             float64
music_genre          object
dtype: object

In [325]:
df.shape

(42791, 14)

In [326]:
df.head()

artist_name  popularity  acousticness  danceability  energy  \
0              Röyksopp        27.0       0.00468         0.652   0.941   
1  Thievery Corporation        31.0       0.01270         0.622   0.890   
2        Dillon Francis        28.0       0.00306         0.620   0.755   
3              Dubloadz        34.0       0.02540         0.774   0.700   
4           What So Not        32.0       0.00465         0.638   0.587   

   instrumentalness key  liveness  loudness   mode  speechiness    tempo  \
0           0.79200  A#     0.115    -5.201  Minor       0.0748  100.889   
1           0.95000   D     0.124    -7.043  Minor       0.0300  115.002   
2           0.01180  G#     0.534    -4.617  Major       0.0345  127.994   
3           0.00253  C#     0.157    -4.498  Major       0.2390  128.014   
4           0.90900  F#     0.157    -6.266  Major       0.0413  145.036   

   valence music_genre  
0    0.759  Electronic  
1    0.531  Electronic  
2    0.333  Electronic  
3    0.270  Electronic  
4    0.323  Electronic

In [327]:
df = df.reset_index(drop=True)

In [328]:
# data variables are object or boolean

cat_cols = list(df.dtypes[df.dtypes == 'object'].index)
print(f'Columns with categorical variables are: {cat_cols} \n')
print(f'Categorical Features: {len(cat_cols)} \n')


num_cols = list(df.dtypes[df.dtypes == 'float64'].index)
print(f'Columns with numerical values are: {num_cols} \n')
print(f'Numerical Features: {len(num_cols)}')

Columns with categorical variables are: ['artist_name', 'key', 'mode', 'music_genre'] 

Categorical Features: 4 

Columns with numerical values are: ['popularity', 'acousticness', 'danceability', 'energy', 'instrumentalness', 'liveness', 'loudness', 'speechiness', 'tempo', 'valence'] 

Numerical Features: 10


## Validation, Testing & Prediction

In [329]:
df.shape

(42791, 14)

##### Replace values in mode to binary

In [330]:
# create a dictionary of replacements
replacements = {'Minor': 0, 'Major': 1}

# replace values using the .map() method
df['mode'] = df['mode'].map(replacements)

#### apply one hot encoding for values in 'Key' column 

In [331]:
ohe = OneHotEncoder(handle_unknown = 'ignore', sparse_output=False).set_output(transform = "pandas")

In [332]:
ohetransform = ohe.fit_transform(df[['key']])

In [333]:
ohetransform.head(3)

key_A  key_A#  key_B  key_C  key_C#  key_D  key_D#  key_E  key_F  key_F#  \
0    0.0     1.0    0.0    0.0     0.0    0.0     0.0    0.0    0.0     0.0   
1    0.0     0.0    0.0    0.0     0.0    1.0     0.0    0.0    0.0     0.0   
2    0.0     0.0    0.0    0.0     0.0    0.0     0.0    0.0    0.0     0.0   

   key_G  key_G#  
0    0.0     0.0  
1    0.0     0.0  
2    0.0     1.0

In [334]:
df = pd.concat([df, ohetransform], axis=1).drop(columns=['key'])

In [335]:
df.head(2)

artist_name  popularity  acousticness  danceability  energy  \
0              Röyksopp        27.0       0.00468         0.652   0.941   
1  Thievery Corporation        31.0       0.01270         0.622   0.890   

   instrumentalness  liveness  loudness  mode  speechiness    tempo  valence  \
0             0.792     0.115    -5.201     0       0.0748  100.889    0.759   
1             0.950     0.124    -7.043     0       0.0300  115.002    0.531   

  music_genre  key_A  key_A#  key_B  key_C  key_C#  key_D  key_D#  key_E  \
0  Electronic    0.0     1.0    0.0    0.0     0.0    0.0     0.0    0.0   
1  Electronic    0.0     0.0    0.0    0.0     0.0    1.0     0.0    0.0   

   key_F  key_F#  key_G  key_G#  
0    0.0     0.0    0.0     0.0  
1    0.0     0.0    0.0     0.0

In [336]:
df.shape

(42791, 25)

In [337]:
df.columns

Index(['artist_name', 'popularity', 'acousticness', 'danceability', 'energy',
       'instrumentalness', 'liveness', 'loudness', 'mode', 'speechiness',
       'tempo', 'valence', 'music_genre', 'key_A', 'key_A#', 'key_B', 'key_C',
       'key_C#', 'key_D', 'key_D#', 'key_E', 'key_F', 'key_F#', 'key_G',
       'key_G#'],
      dtype='object')

In [338]:
# split data set to 60/20/20 for validation and testing
data_full_train, data_test = train_test_split(df, test_size=0.2, random_state=42)
data_train, data_val = train_test_split(data_full_train, test_size=0.25, random_state=42)

In [339]:
len(data_full_train), len(data_test)

(34232, 8559)

In [340]:
len(data_train), len(data_val), len(data_test)

(25674, 8558, 8559)

In [341]:
y_train = data_train.music_genre.values
y_val = data_val.music_genre.values
y_test = data_test.music_genre.values

del data_train['music_genre']
del data_val['music_genre']
del data_test['music_genre']

In [342]:
len(y_train), len(y_val), len(y_test)

(25674, 8558, 8559)

In [343]:
dv = DictVectorizer(sparse=False)

train_dict = data_train.to_dict(orient='records')
X_train = dv.fit_transform(train_dict)

val_dict = data_val.to_dict(orient='records')
X_val = dv.transform(val_dict)

In [344]:
# models to test
classifiers = [
    LogisticRegression(),
    DecisionTreeClassifier(),
    RandomForestClassifier(),
    RidgeClassifier(),
]
# get names of the objects in list (too lazy for c&p...)
names = [re.match(r"[^\(]+", name.__str__())[0] for name in classifiers]
print(f"Classifiers to test: {names}")

Classifiers to test: ['LogisticRegression', 'DecisionTreeClassifier', 'RandomForestClassifier', 'RidgeClassifier']


In [346]:
%%time
# test all classifiers and save pred. results on test data
results = {}
for name, clf in zip(names, classifiers):
    print(f"Training classifier: {name}")
    clf.fit(X_train, y_train)
    prediction = clf.predict(X_val)
    report = classification_report(y_val, prediction)
    results[name] = report

Training classifier: LogisticRegression
Training classifier: DecisionTreeClassifier
Training classifier: RandomForestClassifier
Training classifier: RidgeClassifier
CPU times: total: 11min 32s
Wall time: 7min 37s


In [347]:
# Prediction results
for k, v in results.items():
    print(f"Results for {k}:")
    print(f"{v}\n")

Results for LogisticRegression:
              precision    recall  f1-score   support

 Alternative       0.21      0.21      0.21       858
       Anime       0.46      0.58      0.51       828
       Blues       0.36      0.21      0.26       824
   Classical       0.68      0.79      0.73       818
     Country       0.23      0.23      0.23       879
  Electronic       0.29      0.21      0.24       853
     Hip-Hop       0.34      0.03      0.05       888
        Jazz       0.36      0.41      0.38       879
         Rap       0.31      0.53      0.39       860
        Rock       0.25      0.34      0.29       871

    accuracy                           0.35      8558
   macro avg       0.35      0.35      0.33      8558
weighted avg       0.35      0.35      0.33      8558


Results for DecisionTreeClassifier:
              precision    recall  f1-score   support

 Alternative       0.31      0.30      0.30       858
       Anime       0.73      0.72      0.73       828
       Bl

### Parameters Tuning

#### Logistic Regression

In [234]:
for i in range(200, 250, 10):
    print(f'For {i} \n')
    for n in range(1, 4, 1):
        lr = LogisticRegression(C= n, max_iter= i)

        lr.fit(X_train, y_train)

        lr_score = lr.score(X_val, y_val)

        print(f'C:{n} has an accuracy score: {lr_score} \n')

For 200 

C:1 has an accuracy score: 0.44788501986445434 

C:2 has an accuracy score: 0.434213601308717 

C:3 has an accuracy score: 0.44110773545220844 

For 210 

C:1 has an accuracy score: 0.44893666744566485 

C:2 has an accuracy score: 0.4448469268520682 

C:3 has an accuracy score: 0.44227623276466466 

For 220 

C:1 has an accuracy score: 0.4490535171769105 

C:2 has an accuracy score: 0.4463659733582613 

C:3 has an accuracy score: 0.44367842953961206 

For 230 

C:1 has an accuracy score: 0.4464828230895069 

C:2 has an accuracy score: 0.4467165225519981 

C:3 has an accuracy score: 0.4454311755082963 

For 240 

C:1 has an accuracy score: 0.4482355690581912 

C:2 has an accuracy score: 0.4498714652956298 

C:3 has an accuracy score: 0.4477681701332087 



#### Highest accuracy score of 0.4490 with max_iter = 220 and C=1.0

### Decision Forest

In [237]:
for depth in [5, 10, 15]:
    print(f'For max-depth {depth}: \n')
    for n in range(100, 131, 10):
        model = DecisionTreeClassifier(max_depth=depth, max_leaf_nodes=n, random_state=42)
        model.fit(X_train, y_train)
        rf_score = model.score(X_val, y_val)
        print(f'with max_leaf_nodes, {n} has an accuray_score: {rf_score}')
    print('--------------- \n')

For max-depth 5: 

with max_leaf_nodes, 100 has an accuray_score: 0.4704370179948586
with max_leaf_nodes, 110 has an accuray_score: 0.4704370179948586
with max_leaf_nodes, 120 has an accuray_score: 0.4704370179948586
with max_leaf_nodes, 130 has an accuray_score: 0.4704370179948586
--------------- 

For max-depth 10: 

with max_leaf_nodes, 100 has an accuray_score: 0.5196307548492638
with max_leaf_nodes, 110 has an accuray_score: 0.5220845992054218
with max_leaf_nodes, 120 has an accuray_score: 0.5229025473241412
with max_leaf_nodes, 130 has an accuray_score: 0.5253563916802991
--------------- 

For max-depth 15: 

with max_leaf_nodes, 100 has an accuray_score: 0.5199813040430007
with max_leaf_nodes, 110 has an accuray_score: 0.5212666510867024
with max_leaf_nodes, 120 has an accuray_score: 0.5227856975928955
with max_leaf_nodes, 130 has an accuray_score: 0.5233699462491236
--------------- 



#### Highest accuracy score of 0.5224 with `max_depth = 10 and max_leaf_nodes=130`

### Random Forest

In [239]:
for depth in [10, 15]:
    print(f'For max-depth {depth}: \n')
    for n in range(180, 201, 10):
        model = RandomForestClassifier(n_estimators=n, random_state=42, max_depth = depth)
        model.fit(X_train, y_train)
        rf_score = model.score(X_val, y_val)
        print(f'{n} estimator has an accuray_score: {rf_score}')
    print('--------------- \n')

For max-depth 10: 

180 estimator has an accuray_score: 0.5012853470437018
190 estimator has an accuray_score: 0.4995326010750175
200 estimator has an accuray_score: 0.5014021967749475
--------------- 

For max-depth 15: 

180 estimator has an accuray_score: 0.5337695723299837
190 estimator has an accuray_score: 0.5349380696424398
200 estimator has an accuray_score: 0.5335358728674924
--------------- 



#### Highest accuracy score of 0.5349 with `max_depth = 15 and n_estimator=190`
 Took much time to run without improvement in accuracy

### Ridge Classifier

In [348]:
for i in [0.001, 0.01, 0.1, 1, 10]:
    rcl = RidgeClassifier(alpha=i, random_state=42)

    rcl.fit(X_train, y_train)

    rcl_score = rcl.score(X_val, y_val)

    print(f'for alpha = {i}, accuracy score is {rcl_score}')

for alpha = 0.001, accuracy score is 0.7719093246085534
for alpha = 0.01, accuracy score is 0.7724935732647815
for alpha = 0.1, accuracy score is 0.7755316662771675
for alpha = 1, accuracy score is 0.784295396120589
for alpha = 10, accuracy score is 0.7716756251460621


##### Highest accuracy score of 0.784 with alpha = 1

### The Ridge Classifier provided the highest accuracy score amongst all models 

## K-Fold Validation
For Ridge Classification

In [350]:
len(X_train)

25674

In [351]:
rcl = RidgeClassifier(alpha=1, random_state=42)
accuracies = cross_val_score(rcl, X_train, y_train, cv=10)

print('Acccuracies on each training data:\n', accuracies)
print(' ')
print('Mean:\n', accuracies.mean())

Acccuracies on each training data:
 [0.78037383 0.78777259 0.77686916 0.77647975 0.78184651 0.78340475
 0.775224   0.78496299 0.7693806  0.78262563]
 
Mean:
 0.7798939814831669


#### The Ridge Classification performed across all cros-validated records with an approximate mean of 0.78

## Prediction

In [352]:
dicts_full_train = data_full_train.to_dict(orient='records')

In [353]:
dv = DictVectorizer(sparse=False)
X_full_train = dv.fit_transform(dicts_full_train)

In [354]:
y_full_train = data_full_train.music_genre.values

In [355]:
modelrcl = RidgeClassifier(alpha=1, random_state=42)
modelrcl.fit(X_full_train, y_full_train)

RidgeClassifier(alpha=1, random_state=42)

In [356]:
dicts_test = data_test.to_dict(orient='records')
X_test = dv.transform(dicts_test)

In [358]:
lr_pred = modelrcl.predict(X_test)
lr_class = classification_report(y_test, lr_pred)
lr_score = accuracy_score(y_test, lr_pred)
print(lr_class)
print(f'\n Accuracy score of {np.round(lr_score, 2)} with Ridge Classifier')

              precision    recall  f1-score   support

 Alternative       0.69      0.66      0.68       826
       Anime       0.97      0.99      0.98       867
       Blues       0.84      0.96      0.89       838
   Classical       0.93      0.97      0.95       883
     Country       0.85      0.92      0.88       861
  Electronic       0.91      0.98      0.94       906
     Hip-Hop       0.53      0.66      0.59       850
        Jazz       0.90      0.81      0.85       833
         Rap       0.57      0.43      0.49       840
        Rock       0.75      0.59      0.66       855

    accuracy                           0.80      8559
   macro avg       0.79      0.80      0.79      8559
weighted avg       0.80      0.80      0.79      8559


 Accuracy score of 0.8 with Ridge Classifier


#### Improved result when trained on full data with an accuracy of 0.80

In [361]:
lr_pred

array(['Alternative', 'Alternative', 'Anime', ..., 'Rock', 'Electronic',
       'Classical'], dtype='<U11')

In [359]:
# import pickle

In [360]:
# with open('modelrcl.pkl', 'wb') as f_out:
#     pickle.dump((dv, modelrcl), f_out)

# print(f'the model is saved to modelrcl.pkl')

the model is saved to modelrcl.pkl
